### PR TITLE
Bugfix/log figure correct file extension

### LINKED ISSUE #6039: [BUG] log_figure in plotly format without ".html" extension in artifact_file
### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

1.25.1

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS
- **Python version**: 3.10

### Describe the problem

When logging a figure in plotly format, the file extension in the artifact_file needs to be ".html".
Otherwise, the file will be saved in HTML format anyway, but the file extension will be None or ".png" or other that the user specified in the path string. The file can't be opened on the MLFlow UI later.
The ".html" extension could be added to the end of the path if not there, or the method "write_image" could be used when the path finishes with PNG, JPEG, SVG or PDF.
Cheers,

### Tracking information

_No response_

### Code to reproduce issue

    import mlflow
    import plotly.express as px

    with mlflow.start_run():

        df = px.data.gapminder().query("country=='Canada'")
        fig = px.line(df, x="year", y="lifeExp", title='Life expectancy in Canada')
        mlflow.log_figure(fig, 'fig_test')

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by drsantos89**
## [BUG] log_figure in plotly format without ".html" extension in artifact_file #6039

Closes #6039 

## What changes are proposed in this pull request?

If the figure is in plotly format, the artifact_file string must end with one of the specified extensions. If HTML the write_html method is used. If png, jpeg, webp, SVG or pdf, the write_image is used.

## How is this patch tested?

- [X] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.
tests/tracking/test_log_figure.py
test_log_figure_plotly_png,
test_log_figure_raises_error_for_unsupported_file_extension

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
@drsantos89 Thanks for the PR! Can you remove the unrelated commits?

File: mlflow/tracking/client.py [hunk 1]
```diff
@@ -4,45 +4,55 @@
...
 import contextlib
-import logging
 import json
+import logging
 import os
 import posixpath
 import sys
 import tempfile
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Sequence, Union
+
```
Comment (harupy) [inline mlflow/tracking/client.py:14]: Can you revert these changes? Feel free to open a new PR if necessary.

Comment (drsantos89) [reply mlflow/tracking/client.py:14]: Hi @harupy these changes are the output of isort and don't change anything functionally.
If you think they should not be done here, I will create a new PR
Best,

Comment (drsantos89) [reply mlflow/tracking/client.py]: @harupy I have reverted the changes

File: mlflow/tracking/client.py [hunk 2]
```diff
@@ -1157,7 +1157,13 @@ def _is_plotly_figure(fig):
...
                 figure.savefig(tmp_path)
             elif "plotly" in sys.modules and _is_plotly_figure(figure):
-                figure.write_html(tmp_path, include_plotlyjs="cdn", auto_open=False)
+                file_extension = artifact_file.split('.')[-1]
+                if file_extension == 'html':
+                    figure.write_html(tmp_path, include_plotlyjs="cdn", auto_open=False)
+                elif file_extension in ['png', 'jpeg', 'webp', 'svg', 'pdf']:
+                    figure.write_image(tmp_path)
+                else:
+                    raise TypeError("Unsupported figure extension for plotly")
```
Comment (harupy) [inline mlflow/tracking/client.py:1166]:
```suggestion
                    raise TypeError(f"Unsupported figure extension for plotly: {file_extension}")
```

**Review by harupy — APPROVED** [review]
Left a few comments, otherwise LGTM :)

File: tests/tracking/test_log_figure.py [hunk 3]
```diff
@@ -45,6 +45,36 @@ def test_log_figure_plotly(subdir):
         assert os.listdir(run_artifact_dir) == [filename]
 
 
+def test_log_figure_plotly_png():
```
Comment (harupy) [inline tests/tracking/test_log_figure.py:48]:
```suggestion
@pytest.mark.parametrize("extension", ['png', 'jpeg', 'webp', 'svg', 'pdf'])
def test_log_figure_plotly_image(extension):
```

File: tests/tracking/test_log_figure.py [hunk 4]
```diff
@@ -45,6 +45,36 @@ def test_log_figure_plotly(subdir):
...
+    from plotly import graph_objects as go
+    
+    subdir = "."
+    filename = "figure.png"
```
Comment (harupy) [inline tests/tracking/test_log_figure.py:52]:
```suggestion
    filename = f"figure.{extension}"
```

File: tests/tracking/test_log_figure.py [hunk 5]
```diff
@@ -45,6 +45,36 @@ def test_log_figure_plotly(subdir):
...
+
+def test_log_figure_raises_error_for_unsupported_file_extension():
+    from plotly import graph_objects as go
+
+    filename = "figure"
+    artifact_file = posixpath.join(".", filename)
+
+    fig = go.Figure(go.Scatter(x=[0, 1], y=[2, 3]))
+
+    with mlflow.start_run(), pytest.raises(TypeError, match="Unsupported figure extension for plotly"):
```
Comment (drsantos89) [inline tests/tracking/test_log_figure.py:74]:
```suggestion
    with mlflow.start_run(), pytest.raises(TypeError, match=f"Unsupported file extension for plotly figure: figure"):
```

File: mlflow/tracking/client.py [hunk 6]
```diff
@@ -1157,7 +1157,13 @@ def _is_plotly_figure(fig):
             if "matplotlib" in sys.modules and _is_matplotlib_figure(figure):
                 figure.savefig(tmp_path)
             elif "plotly" in sys.modules and _is_plotly_figure(figure):
-                figure.write_html(tmp_path, include_plotlyjs="cdn", auto_open=False)
+                file_extension = artifact_file.split('.')[-1]
```
Comment (harupy) [inline mlflow/tracking/client.py:1160]:
```suggestion
                file_extension = os.path.splitext(artifact_file)[1]
```

File: tests/tracking/test_log_figure.py [hunk 7]
```diff
@@ -45,6 +45,37 @@ def test_log_figure_plotly(subdir):
...
+
+def test_log_figure_raises_error_for_unsupported_file_extension():
+    from plotly import graph_objects as go
+
+    filename = "figure"
+    artifact_file = posixpath.join(".", filename)
+
+    fig = go.Figure(go.Scatter(x=[0, 1], y=[2, 3]))
+
+    with mlflow.start_run(), pytest.raises(TypeError, match=f"Unsupported file extension for plotly figure: ''"):
```
Comment (harupy) [inline tests/tracking/test_log_figure.py:75]:
```suggestion
    with mlflow.start_run(), pytest.raises(TypeError, match="Unsupported file extension for plotly figure: ''"):
```

File: tests/tracking/test_log_figure.py [hunk 8]
```diff
@@ -45,6 +45,37 @@ def test_log_figure_plotly(subdir):
...
+    with mlflow.start_run():
+        mlflow.log_figure(fig, artifact_file)
+
+        artifact_path = None if subdir is None else posixpath.normpath(subdir)
+        artifact_uri = mlflow.get_artifact_uri(artifact_path)
+        run_artifact_dir = local_file_uri_to_path(artifact_uri)
+        assert os.listdir(run_artifact_dir) == [filename]     
+        
+
+def test_log_figure_raises_error_for_unsupported_file_extension():
```
Comment (harupy) [inline tests/tracking/test_log_figure.py:67]:
```suggestion
@pytest.mark.parametrize("extension", ['', '.py'])
def test_log_figure_raises_error_for_unsupported_file_extension(extension):
```

Can we also parametrize this test?

File: tests/tracking/test_log_figure.py [hunk 9]
```diff
@@ -45,6 +45,37 @@ def test_log_figure_plotly(subdir):
...
+    from plotly import graph_objects as go
+
+    filename = "figure"
```
Comment (harupy) [inline tests/tracking/test_log_figure.py:70]:
```suggestion
    filename = f"figure{extension}"
```

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/runs/6882541571?check_suite_focus=true#step:9:1336

```
E           ValueError: 
E           Image export using the "kaleido" engine requires the kaleido package,
E           which can be installed using pip:
E               $ pip install -U kaleido
```

It seems we need to install `kaleido`.

**Comment by harupy** [comment]
autoformat

**Pull request merged**